makeflow/src/mountfile.c

```diff
@@ -0,0 +1,660 @@
+/*
+Copyright (C) 2008- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+#include <assert.h>
+#include <errno.h>
+#include <libgen.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "copy_tree.h"
+#include "copy_stream.h"
+#include "create_dir.h"
+#include "debug.h"
+#include "http_query.h"
+#include "list.h"
+#include "makeflow_log.h"
+#include "md5.h"
+#include "mkdir_recursive.h"
+#include "mountfile.h"
+#include "path.h"
+#include "unlink_recursive.h"
+#include "xxmalloc.h"
+
+#define HTTP_TIMEOUT 300
+
+/* create_link creates a link from link_name to link_target.
+ * first try to create a hard link, then try to create a symlink when failed to create a hard link.
+ * return 0 on success, non-zero on failure.
+ */
+int create_link(const char *link_target, const char *link_name) {
+	if(link(link_target, link_name)) {
+		LDEBUG("link(%s, %s) failed: %s!\n", link_target, link_name, strerror(errno));
+		if(symlink(link_target, link_name)) {
+			LDEBUG("symlink(%s, %s) failed: %s!\n", link_target, link_name, strerror(errno));
+			return -1;
+		}
+	}
+	return 0;
+}
+
+/* md5_cal calculates the md5 checksum of string s.
+ * @param s: a string pointer
+ * return the md5 checksum of s on success, return NULL on failure.
+ * The caller should free the returned string.
+ */
+char *md5_cal(const char *s) {
```

Comment:
We seem to do this often. Could you move this one to md5.* in dttools (and name it md5_of_string?)